ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp

```diff
@@ -0,0 +1,168 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "unary_device_operation.hpp"
+#include "third_party/magic_enum/magic_enum.hpp"
+#include "ttnn/tensor/tensor_utils.hpp"
+#include "tt_metal/common/constants.hpp"
+#include "tt_metal/host_api.hpp"
+#include "tt_metal/tools/profiler/op_profiler.hpp"
+#include "ttnn/operations/eltwise/unary/common/unary_op_utils.hpp"
+
+namespace ttnn::operations::unary {
+
+namespace {
+void validate_supported_arch_dtype(tt::ARCH arch, DataType input_datatype, DataType output_datatype, UnaryOpType op_type) {
+    switch (op_type) {
+        case UnaryOpType::REMAINDER:
+        case UnaryOpType::FLOOR:
+        case UnaryOpType::CEIL:
+        case UnaryOpType::LEFT_SHIFT:
+        case UnaryOpType::RIGHT_SHIFT:
+            TT_FATAL(arch != tt::ARCH::GRAYSKULL, fmt::format("UnaryOpType '{}' is not supported on Grayskull architecture.", static_cast<int>(op_type)));
+            break;
+        case UnaryOpType::BITWISE_XOR:
+        case UnaryOpType::BITWISE_NOT:
+        case UnaryOpType::BITWISE_AND:
+        case UnaryOpType::BITWISE_OR:
+            TT_FATAL(arch != tt::ARCH::GRAYSKULL, fmt::format("UnaryOpType '{}' is not supported on Grayskull architecture (Bitwise operation).", static_cast<int>(op_type)));
+            TT_FATAL(input_datatype == DataType::INT32, fmt::format("Unsupported input data type '{}' for UnaryOpType '{}' (Bitwise operation).", static_cast<int>(input_datatype), static_cast<int>(op_type)));
+            TT_FATAL(output_datatype == DataType::INT32, fmt::format("Unsupported output data type '{}' for UnaryOpType '{}' (Bitwise operation).", static_cast<int>(output_datatype), static_cast<int>(op_type)));
+            break;
+        case UnaryOpType::FMOD:
+            TT_FATAL(arch != tt::ARCH::GRAYSKULL, fmt::format("UnaryOpType '{}' (FMOD operation) is not supported on Grayskull architecture.", static_cast<int>(op_type)));
+            TT_FATAL(input_datatype == DataType::BFLOAT16, fmt::format("Unsupported input data type '{}' for UnaryOpType '{}' (FMOD operation).", static_cast<int>(input_datatype), static_cast<int>(op_type)));
+            TT_FATAL(output_datatype == DataType::BFLOAT16, fmt::format("Unsupported output data type '{}' for UnaryOpType '{}' (FMOD operation).", static_cast<int>(output_datatype), static_cast<int>(op_type)));
+            break;
+        default:
+            return;
+    }
+}
+} // namespace
+
+
+UnaryDeviceOperation::program_factory_t UnaryDeviceOperation::select_program_factory(const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    if (tensor_args.input.is_sharded()) {
+        return program::UnaryShardedProgramFactory{};
+    }
+    else {
+        return program::UnaryProgramFactory{};
+    }
+}
+
+void UnaryDeviceOperation::validate_on_program_cache_hit(const operation_attributes_t& args, const tensor_args_t& tensor_args) {
```

Comment:
@arakhmati , is this ok?

Reply:
Yes. Re-using `validate_on_program_cache_miss` in `validate_on_program_cache_hit` is totally fine in my opinion. They only need to be different if the op writer wants them to be different